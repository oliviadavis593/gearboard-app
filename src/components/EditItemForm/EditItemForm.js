import React, { Component} from 'react';
import GearContext from '../../GearContext'
import MainHeader from '../../views/MainHeader';
import TokenService from '../../services/token-service'
import config from '../../config';
import '../../styles/EditItem.css'

export default class EditItemForm extends Component {
    state = {
        rating: '',
        features: '',
        comment: '',
        gear_name: '',
        error: null
    }

    static defaultProps =  {
        history: {
            goBack: () => {}
        },
        match: {
            params: {}
        }
    }

    static contextType = GearContext; 

    componentDidMount() {
        const item_id = this.props.match.params.item_id
        console.log("item_id", item_id)
        fetch(`${config.API_ENDPOINT}/api/items/${item_id}`, {
            method: 'GET',
            headers: {
                'authorization': `basic ${TokenService.getAuthToken()}`
            }
        })
        .then(res => {
            if (!res.ok) {
                throw new Error(res.status)
            }
            return res.json()
        })
        .then(responseData => {
            this.setState({ 
                id: responseData.id,
                rating: responseData.rating, 
                gear_name: responseData.gear_name, 
                features: responseData.features, 
                comments: responseData.comments
            })
        })
        .catch(error => {
            console.error(error)
            this.setState({ error })
        })
    }

    handleChangeGearName = e => {
        this.setState({ gear_name: e.target.value })
    }

    handleChangeDescription = e => {
        this.setState({ features: e.target.value })
    }

    handleChangeComment = e => {
        this.setState({ comments: e.target.value })
    }

    handleChangeRating = e => {
        this.setState({ rating: e.target.value })
    }

    handleSubmit = e => {
        e.preventDefault()
        const { item_id } = this.props.match.params
        const { id, rating, gear_name, features, comments } = this.state
        const newItem = { id, rating, gear_name, features, comments } 
        fetch(`${config.API_ENDPOINT}/api/items/${item_id}`, {
            method: 'PATCH',
            body: JSON.stringify(newItem),
            headers: {
                'content-type': 'application/json'
            }
        })
        .then(res => {
            if (!res.ok) {
                throw new Error(res.status)
            }
            //return res.json()
        })
        .then(res => {
            this.context.updateItem(newItem)
            this.props.history.push('/homepage')
        })
        .catch(error => {
            console.error({ error })
            this.setState({ error })
        })
    }

    render() {
        
        return(
            <div>
                <header>
                    <MainHeader />
                </header>
                <main>
                    <form 
                    className='gb-edit-item__form'
                    onSubmit={this.handleSubmit}
                    >
                        <fieldset>
                            <legend><h2>Edit Gear Form</h2></legend>
                            <label htmlFor='gear-name'>Gear Name:</label>
                            <input 
                            id='gear-name'
                            type='text'
                            name='gear_name'
                            required
                            onChange={this.handleChangeGearName}
                            />
    
                            <label htmlFor='description'>Features:</label>
                            <textarea
                            id='features'
                            type='textarea'
                            name='features'
                            required
                            onChange={this.handleChangeDescription}
                            >
                            </textarea>
    
                            <label htmlFor='comment'>Comments:</label>
                            <textarea
                            id='comment'
                            type='textarea'
                            name='comments'
                            onChange={this.handleChangeComment}
                            >
                            </textarea>
                        
                        <label htmlFor='rating'>Rating:</label>
                        <select name='rating' id='gb-rating' onChange={this.handleChangeRating}>
                            <option value=''></option>
                            <option value='ðŸŽ¸'>ðŸŽ¸</option>
                            <option value='ðŸŽ¸ðŸŽ¸'>ðŸŽ¸ðŸŽ¸</option>
                            <option value='ðŸŽ¸ðŸŽ¸ðŸŽ¸'>ðŸŽ¸ðŸŽ¸ðŸŽ¸</option>
                            <option value='ðŸŽ¸ðŸŽ¸ðŸŽ¸ðŸŽ¸'>ðŸŽ¸ðŸŽ¸ðŸŽ¸ðŸŽ¸</option>
                            <option value='ðŸŽ¸ðŸŽ¸ðŸŽ¸ðŸŽ¸ðŸŽ¸'>ðŸŽ¸ðŸŽ¸ðŸŽ¸ðŸŽ¸ðŸŽ¸</option>
                        </select>
    
                        </fieldset>
    
                        <button
                        className='gb-edit__button'
                        >
                            Edit Gear
                        </button>
                    </form>
                    <div>
                        <button
                        tag='button'
                        role='link'
                        className='gb-edit__goback-button'
                        onClick={() => this.props.history.goBack()} 
                        >
                            Back
                        </button>
                    </div>
                </main>
            </div>
        )
    }
}