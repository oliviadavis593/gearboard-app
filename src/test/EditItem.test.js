import React from 'react';
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import EditItemForm from '../components/EditItemForm/EditItemForm'

describe('EditItem component', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div')
        ReactDOM.render(<EditItemForm />, div)
        ReactDOM.unmountComponentAtNode(div)
    })

    it('renders the UI as expected', () => {
        const tree = renderer 
            .create(<EditItemForm gear_name='test gear_name' features='test features' comments='test comments' rating='ðŸŽ¸' />)
            .toJSON()
           expect(tree).toMatchSnapshot();
    })
})