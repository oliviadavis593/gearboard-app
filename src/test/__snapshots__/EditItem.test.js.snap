// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditItem component renders the UI as expected 1`] = `
<div>
  <header>
    <div
      className="gb-header__container"
    >
      <header
        className="gb-main__header"
      >
        <a
          aria-current={null}
          href="/homepage"
          onClick={[Function]}
        >
          <h1>
            GearBoard
          </h1>
        </a>
      </header>
      <nav
        className="nav"
      >
        <div
          className="nav__main-links"
        >
          <a
            aria-current={null}
            className="gb-link__all"
            href="/homepage"
            onClick={[Function]}
          >
            All Gear
          </a>
          <a
            aria-current={null}
            className="gb-link__new"
            href="/new-item"
            onClick={[Function]}
          >
            Submit new item
          </a>
          <a
            aria-current="page"
            className="gb-link__logout active"
            href="/"
            onClick={[Function]}
            style={Object {}}
          >
            Logout
          </a>
        </div>
        <span
          className="navbar-toggle"
          id="js-navbar-toggle"
        />
      </nav>
    </div>
  </header>
  <main>
    <form
      className="gb-edit-item__form"
      onSubmit={[Function]}
    >
      <fieldset>
        <legend>
          <h2>
            Edit Gear Form
          </h2>
        </legend>
        <label
          htmlFor="gear-name"
        >
          Gear Name:
        </label>
        <input
          id="gear-name"
          name="gear_name"
          onChange={[Function]}
          required={true}
          type="text"
        />
        <label
          htmlFor="description"
        >
          Features:
        </label>
        <textarea
          id="features"
          name="features"
          onChange={[Function]}
          required={true}
          type="textarea"
        />
        <label
          htmlFor="comment"
        >
          Comments:
        </label>
        <textarea
          id="comment"
          name="comments"
          onChange={[Function]}
          type="textarea"
        />
        <label
          htmlFor="rating"
        >
          Rating:
        </label>
        <select
          id="gb-rating"
          name="rating"
          onChange={[Function]}
        >
          <option
            value=""
          />
          <option
            value="ðŸŽ¸"
          >
            ðŸŽ¸
          </option>
          <option
            value="ðŸŽ¸ðŸŽ¸"
          >
            ðŸŽ¸ðŸŽ¸
          </option>
          <option
            value="ðŸŽ¸ðŸŽ¸ðŸŽ¸"
          >
            ðŸŽ¸ðŸŽ¸ðŸŽ¸
          </option>
          <option
            value="ðŸŽ¸ðŸŽ¸ðŸŽ¸ðŸŽ¸"
          >
            ðŸŽ¸ðŸŽ¸ðŸŽ¸ðŸŽ¸
          </option>
          <option
            value="ðŸŽ¸ðŸŽ¸ðŸŽ¸ðŸŽ¸ðŸŽ¸"
          >
            ðŸŽ¸ðŸŽ¸ðŸŽ¸ðŸŽ¸ðŸŽ¸
          </option>
        </select>
      </fieldset>
      <button
        className="gb-edit__button"
      >
        Edit Gear
      </button>
    </form>
    <div>
      <button
        className="gb-edit__goback-button"
        onClick={[Function]}
        role="link"
        tag="button"
      >
        Back
      </button>
    </div>
  </main>
</div>
`;
